### PR TITLE
Bowling Challenge - Complete

### PR DESCRIPTION
**Pull request by nyahehnagi**
I'd like to improve on how I calculate bonus points. Pretty sure it can be further refactored..  I am still working on it

I've incorporated a LastFrame as an inherited class from Frame. This was for 2 reasons, I wanted to see how `super ` works and this seemed a good opportunity to see it in action, plus the last frame in bowling exhibed additional behaviour over a normal frame. Now.. I've read more.. it seems I can modules.. ok.. off I go to read about this and see if there's a refactor needed :)

I have intentionally not mocked the Frame and LastFrame classes in Scorecard. Although I have passed in these dependancies into Scorecard, when it comes to functionality I feel these classes are always going to be together. I cannot see how the Frame class could be used elsewhere but the Scorecard.  Adding mocks would a) be very time consuming for little value.. the tests for the Scorecard cover mulitple complex scenarios where the input from the Frame classes is key to score evaluation b) adding mocks would inhibit the refactoring process on Scorecard. I'd be keen to discuss this with a coach on their thoughts.. it could be my design is well off and to complex.. and that could be a warning sign to the design approach I took

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,65 +1,28 @@
...
+~~~~
+git clone git@github.com:nyahehnagi/bowling-challenge-ruby.git
+~~~~
+
+## Testing
+-----------------------
+To run rspec tests
+~~~~
+$> rspec
+~~~~
```
Comment (winters145) [inline README.md:28]: Maybe some info or instruction on how to run the program would be helpful

File: lib/frame.rb [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+  def invalid_pins?(pins_downed)
+    pins_downed.negative? || pins_downed > TEN_PINS
+  end
+
+  def to_many_pins?(pins_downed)
+    return false if @frame_type == :strike
+
+    pins_downed > pins_left
+  end
```
Comment (winters145) [inline lib/frame.rb:59]: Lots of edge cases covered

File: lib/frame.rb [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    @rolls << pins_downed
+  end
+
+  def process_frame_type(pins_downed)
+    return unless @frame_type.nil?
+
+    @frame_type = :strike if pins_downed == TEN_PINS && first_roll?
+    @frame_type = :spare if second_roll? && pins_downed + @rolls[FIRST_ROLL] == TEN_PINS
+    @frame_type = :open if second_roll? && pins_downed + @rolls[FIRST_ROLL] != TEN_PINS
```
Comment (winters145) [inline lib/frame.rb:88]: I like the setting of @frame_type

File: lib/last_frame.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+require './lib/frame'
+
+class LastFrame < Frame
```
Comment (winters145) [inline lib/last_frame.rb:5]: Interesting use of a sub-class, impressive! (Although I'm finding it quite hard to follow exactly what's happening!)

File: lib/frame.rb [hunk 5]
```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+
+class Frame
+  TEN_PINS = 10
+  FIRST_ROLL = 0
+  SECOND_ROLL = 1
```
Comment (winters145) [inline lib/frame.rb:6]: Nice use of constants to keep the code clear and obvious as to what is being referenced

File: lib/scorecard.rb [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    frame.strike_frame? ? STRIKE : SPARE
+  end
+
+  def bonus_points(frame_no, rolls_ahead)
+    frames_ahead = @frames.select { |frame_key, _frame| frame_key >= frame_no }
+    
+    frame_scores = frames_ahead.map { |_frame_key, frame| frame.all_rolls }.flatten
+
+    frame_scores = rolls_ahead == SPARE ? frame_scores.drop(2) : frame_scores.drop(1)
+    frame_scores.first(rolls_ahead).count == rolls_ahead ? frame_scores.first(rolls_ahead).sum : ZERO_SCORE
```
Comment (winters145) [inline lib/scorecard.rb:79]: Clever code, but quite hard to read / understand what is actually being done here

File: lib/scorecard.rb [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    if next_frame?
+      @frames[@frames.count + 1] = last_frame_next? ? @last_frame_class.new : @frame_class.new
+    end
+    @frames[@frames.keys.last]
+  end
+
+  def next_frame?
+    return true if @frames.empty?
+    return false if @frames.count == LAST_FRAME
+    return true if @frames[@frames.keys.last].frame_complete?
```
Comment (winters145) [inline lib/scorecard.rb:59]: Clever bit of iteration over the frames hash

File: spec/frame_spec.rb [hunk 8]
```diff
@@ -0,0 +1,138 @@
+# frozen_string_literal: true
```
Comment (winters145) [inline spec/frame_spec.rb:1]: Incredibly wide-ranging tests, great coverage

**Review by winters145 — COMMENTED** [review]
Great code! Really well extracted methods and classes - SRP followed to a tee! Very in-depth testing too, all possible edge cases seem to have been covered, good job!

File: lib/frame.rb [hunk 9]
```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
```
Comment (nelsonclaire) [inline lib/frame.rb:1]: Love the layout of the classes, easy to read and understandable.

File: README.md [hunk 10]
```diff
@@ -1,65 +1,28 @@
...
+----------------------
+## Installation
+----------------------
+Install gems with [bundler](https://bundler.io/ "bundler") 
+~~~~
+$> bundle install 
+~~~~
+To install code from the latest source
+~~~~
+git clone git@github.com:nyahehnagi/bowling-challenge-ruby.git
```
Comment (nelsonclaire) [inline README.md:20]: is this correct (thinking it is git@github.com/nyahehnagi/bowling-challenge-ruby.git)?

File: README.md [hunk 11]
```diff
@@ -1,65 +1,28 @@
...
-### Perfect Game
-
-A Perfect Game is when the player rolls 12 strikes (10 regular strikes and 2 strikes for the bonus in the 10th frame). The Perfect Game scores 300 points.
-
-In the image below you can find some score examples.
-
-More about ten pin bowling here: http://en.wikipedia.org/wiki/Ten-pin_bowling
-
-![Ten Pin Score Example](images/example_ten_pin_scoring.png)
+# Bowling Challenge
```
Comment (nelsonclaire) [inline README.md:1]: Did you consider doing any modelling before starting the code and capturing it?

File: lib/last_frame.rb [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+  LAST_ROLL = 2
+  
+  def frame_complete?
+    return @rolls.count == 3 if bonus_frame?
+
+    super
+  end
+
+  def log_roll(pins_downed)
```
Comment (nelsonclaire) [inline lib/last_frame.rb:14]: Good clear error messages

File: lib/scorecard.rb [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+  end
+
+  def strike_or_spare(frame)
+    frame.strike_frame? ? STRIKE : SPARE
+  end
+
+  def bonus_points(frame_no, rolls_ahead)
+    frames_ahead = @frames.select { |frame_key, _frame| frame_key >= frame_no }
+    
+    frame_scores = frames_ahead.map { |_frame_key, frame| frame.all_rolls }.flatten
```
Comment (nelsonclaire) [inline lib/scorecard.rb:76]: Neat - now need to look up what these do!

File: spec/features/logging_rolls_to_scorcard_spec.rb [hunk 14]
```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+describe 'logging rolls to the scorecard' do
+  it 'first frame, first roll, no strike' do
+    scorecard = ScoreCard.new
```
Comment (nelsonclaire) [inline spec/features/logging_rolls_to_scorcard_spec.rb:5]: Are these needed in every test?

File: spec/frame_spec.rb [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+
+    it 'shows a frame is not a spare frame after an open frame' do
+      frame.log_roll(7)
+      frame.log_roll(1)
+      expect(frame.spare_frame?).to be false
+    end
+  end
+
+  context 'input validation' do
+    it 'does not allow rolls less than 0' do
```
Comment (nelsonclaire) [inline spec/frame_spec.rb:91]: Could this actually be input?

File: lib/frame.rb [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+    process_roll(pins_downed)
+  end
+
+  def sum_frame
+    @rolls.sum
+  end
+
+  def all_rolls
+    @rolls
+  end
```
Comment (JohnForster) [inline lib/frame.rb:27]: This method seems unnecessary to me, as the rolls can be accessed directly through the `@rolls` instance method.

File: lib/frame.rb [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+    @frame_type == :strike
+  end
+
+  def spare_frame?
+    @frame_type == :spare
+  end
+
+  def open_frame?
+    @frame_type == :open
+  end
```
Comment (JohnForster) [inline lib/frame.rb:43]: Ideally we want to minimise the amount of state stored in a class. Whether the frame is a strike or spare can be calculated directly from other state (the rolls), so it's best not to duplicate it in a `@frame_type` variable.

Comment (JohnForster) [reply lib/frame.rb:43]: Instead, consider a strike? method which works out from the rolls whether or not the frame is a strike.

File: lib/frame.rb [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+
+  private
+
+  def invalid_roll?(roll)
+    roll.negative? || (roll > @rolls.count)
+  end
```
Comment (JohnForster) [inline lib/frame.rb:49]: I don't think this method is being used.

File: lib/frame.rb [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+    TEN_PINS - @rolls[FIRST_ROLL]
+  end
+
+  def process_roll(pins_downed)
+    process_second_roll(pins_downed)
+    process_first_roll(pins_downed)
+  end
+
+  def process_second_roll(pins_downed)
+    return unless second_roll?
```
Comment (JohnForster) [inline lib/frame.rb:71]: It would be better to move this conditional into the `process_roll` method, so that you're only ever calling `process_second_roll` when you are processing a second roll.

Comment (JohnForster) [reply lib/frame.rb:71]: Same goes for the `return unless first_roll?` below

File: lib/frame.rb [hunk 20]
```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+
+class Frame
+  TEN_PINS = 10
```
Comment (JohnForster) [inline lib/frame.rb:4]: Better to name this constant NUMBER_OF_PINS, because TEN_PINS is just as inflexible as using the number 10. If I asked you to adapt your program for a version of bowling that had 15 pins instead of 10, for example, this constant name wouldn't make sense any more.

File: lib/frame.rb (hunk 2 above)
Comment (JohnForster) [inline lib/frame.rb:59]: I don't see a reason to make this method dependent on the frame_type. The logic should work whether or not the frame type was a strike.

File: lib/frame.rb [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+    @frame_type = :strike if pins_downed == TEN_PINS && first_roll?
+    @frame_type = :spare if second_roll? && pins_downed + @rolls[FIRST_ROLL] == TEN_PINS
+    @frame_type = :open if second_roll? && pins_downed + @rolls[FIRST_ROLL] != TEN_PINS
+  end
+
+  def second_roll?
+    return false if first_roll?
+
+    @rolls[SECOND_ROLL].nil?
+  end
```
Comment (JohnForster) [inline lib/frame.rb:95]: If you use the .length property of the `@rolls` away, this method and others can become simpler.

File: lib/last_frame.rb (hunk 4 above)
Comment (JohnForster) [inline lib/last_frame.rb:5]: Nice idea to use an inherited class for the last frame!

File: lib/last_frame.rb [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+    raise 'Pins downed must be between 0 and 10' if invalid_pins?(pins_downed)
+    raise 'Frame complete. Cannot roll again' if frame_complete?
+
+    process_frame_type(pins_downed)
+    process_roll(pins_downed)
+  end
```
Comment (JohnForster) [inline lib/last_frame.rb:20]: As far as I can see this method is the same as in the Frame class, so shouldn't be duplicated.

File: lib/scorecard.rb [hunk 23]
```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+require 'frame'
+
+class ScoreCard
+  LAST_FRAME = 10
+  TEN_PINS = 10
+  ZERO_SCORE = 0
```
Comment (JohnForster) [inline lib/scorecard.rb:8]: This doesn't need to be a constant.

File: lib/scorecard.rb [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+  SPARE = 1
+  STRIKE = 2
+
+  def initialize(frame_class = Frame, last_frame_class = LastFrame)
```
Comment (JohnForster) [inline lib/scorecard.rb:12]: Good dependency injection!

File: lib/scorecard.rb [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+    @last_frame_class = last_frame_class
+    @frames = {}
+  end
+
+  def current_frame_number
+    @frames.count
+  end
+
+  def log_roll(pins_downed)
+    return :game_complete if game_complete?
```
Comment (JohnForster) [inline lib/scorecard.rb:23]: Trying to enter a roll to a completed game should probably raise an exception.

File: lib/scorecard.rb (hunk 6 above)
Comment (JohnForster) [reply lib/scorecard.rb:79]: This could be improved by refactoring into clear methods.

File: spec/features/logging_rolls_to_scorcard_spec.rb [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  it 'having a bad day.. Gutter Game!' do
+    scorecard = ScoreCard.new
+    20.times do
+      scorecard.log_roll(0)
+    end
+
+    expect(scorecard.score).to eq 0
+    expect(scorecard.current_frame_number).to eq 10
```
Comment (JohnForster) [inline spec/features/logging_rolls_to_scorcard_spec.rb:49]: Excellent feature tests.

**Review by JohnForster — COMMENTED** [review]
This is generally an excellent solution Bromley, well done. There are a few areas that could be simplified with a bit of refactoring, and currently the logic for calculating the score of a frame is split between the Frame class and the Scorecard class. Unifying this might simplify the logic. 

RE: Your question about mocking, the reason we mock our tests is to make our code more flexible. Lets say we were using the logic you have written for an online bowling game. If we wanted to add a mode where there were 15 pins insead of 10, how easy would that addition be to make?

As it stands, the tests for your Scorecard class is very much reliant on the frame class we've already written, and we'd need to change a lot of tests to make sure that it would also work with our new expanded frame class. In a perfect world, we would be able to write a new ExpandedFrame class, and be confident that it would work with the existing Scorecard. Mocking the Frame class helps to give us this confidence.

File: lib/frame.rb (hunk 17 above)
Comment (nyahehnagi) [reply lib/frame.rb:43]: Implemented in the javascript version. Thanks for the feedback

File: lib/frame.rb (hunk 19 above)
Comment (nyahehnagi) [reply lib/frame.rb:71]: implemented in the javascript version.  Although not finished I don't think I need to concern myself about 1st and second rolls now as I can now just look at the isStrike/Spare methods
